connection.go

```diff
@@ -0,0 +1,409 @@
+					memphisKafkaErr(err.Error())
+				}
+
+				_, err = nc.Request(clientReconnectionUpdateSubject, clientReconnectionUpdateReqBytes, 5*time.Second)
+				if err != nil {
+					memphisKafkaErr(err.Error())
+				}
+
+				c.NatsConnectionID = natsConnectionID
+			},
+		),
+	}
+
+	nc, err := nats.Connect(host, opts...)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+
+	js, err := nc.JetStream()
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+
+	c.BrokerConnection = nc
+	c.JSContext = js
+
+	natsConnectionID, err := c.generateNatsConnectionID()
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+	c.NatsConnectionID = natsConnectionID
+
+	return nil
+}
+
+func (c *Client) RegisterClient() error {
+	registerReq := RegisterReq{
+		NatsConnectionID: c.NatsConnectionID,
+		Language:         "go",
+	}
+
+	registerReqBytes, err := json.Marshal(registerReq)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	resp, err := c.BrokerConnection.Request(clientRegisterSubject, registerReqBytes, 5*time.Second)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	var registerResp RegisterResp
+	err = json.Unmarshal(resp.Data, &registerResp)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	c.ClientID = registerResp.ClientID
+	c.AccountName = registerResp.AccountName
+	c.LearningFactor = registerResp.LearningFactor
+	c.LearningFactorCounter = 0
+	c.LearningRequestSent = false
+	c.ConsumerProtoDescMap = make(map[string]protoreflect.MessageDescriptor)
+	c.IsConsumer = false
+	c.IsProducer = false
+
+	return nil
+}
+
+func (c *Client) SubscribeUpdates() error {
+	cus := ClientUpdateSub{
+		ClientID:   c.ClientID,
+		UpdateCahn: make(chan Update),
+	}
+
+	go cus.UpdatesHandler()
+
+	var err error
+	cus.Subscription, err = c.BrokerConnection.Subscribe(fmt.Sprintf(memphisClientUpdatesSubject, c.ClientID), cus.SubscriptionHandler())
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+
+	return nil
+}
+
+func (c *ClientUpdateSub) UpdatesHandler() {
+	for {
+		msg := <-c.UpdateCahn
+		switch msg.Type {
+		case "LearnedSchema":
+			desc := compileMsgDescriptor(msg.Payload)
+			if desc != nil {
+				ClientConnection.ProducerProtoDesc = desc
+			}
+		case "SchemaUpdate":
+			desc := compileMsgDescriptor(msg.Payload)
+			var update SchemaUpdateReq
+			err := json.Unmarshal(msg.Payload, &update)
+			if err != nil {
+				memphisKafkaErr(err.Error())
+			}
+			if desc != nil {
+				ClientConnection.ConsumerProtoDescMap[update.SchemaID] = desc
+			}
+
+		}
+	}
+}
+
+func (c *ClientUpdateSub) SubscriptionHandler() nats.MsgHandler {
+	return func(msg *nats.Msg) {
+		var update Update
+		err := json.Unmarshal(msg.Data, &update)
+		if err != nil {
+			memphisKafkaErr(err.Error())
```

Comment:
add an indication from where the error comes so we will be able to understand where to go and check

make sure it is happening in anywhere we are sending the error to us.
And it will be good to add the SDK lang + version to this message